### DISCUSSION TITLE
RabbitMQ with MQTT - memory usage growing over time

### DISCUSSION THREAD

**Discussion by kbudde** [rabbitmq/rabbitmq-server#3164] [Questions]
Hello,

I've followed the tickets and discussions quite carefully over the last monthes and I've seen many improvements so far but still it looks like the memory in use is growing over time until we reach the high watermark.

I'm looking for help what needs to be optimized or maybe there is still a bug left?

**Environment**
Ubuntu 18.04 with 4 CPUs, 16GB Memory
RabbitMQ Version: 3.8.14 # RabbitMQ was installed using apt.
Erlang Version: 23.3.4

We have around 11k devices connecting with mqtt protocol but not more than around 5k at the same time.
We have a "predictable workload". It's higher during office hours (highlighted in grey) and lower than <500 on weekends. In green you can see average memory usage per hour. (Peak memory usage is higher)
![image](https://user-images.githubusercontent.com/1072181/123960141-3ebcd900-d9af-11eb-8a44-380effaa47b7.png)

I've just checked the current memory details
![image](https://user-images.githubusercontent.com/1072181/123960634-d02c4b00-d9af-11eb-8dde-32b9aee15ed0.png)

and the values from today and 21 days ago with a modified "erlang memory-Allocators" dashboard
![image](https://user-images.githubusercontent.com/1072181/123962399-b7bd3000-d9b1-11eb-9c85-036020a7c6e6.png)

I see it increasing but I don't know why.

At the moment we restart the service every month and it's always the same. It starts low and grows over time.
It would be very nice if someone could suggest next steps. 
I suspect we have to tune the memory settings, but it's a production system; so I would be happy about some hints from people with more erlang experience.
https://www.rabbitmq.com/runtime.html#allocators

**Comment by lhoguin** [comment]
You have a large increase in the `ets_alloc`. This suggests that some data might be accumulating in an ets table or fragmentation.

`eheap_alloc` only has a small increase. I don't know if that's "normal" or not.

Based on the information here it's difficult to know. Knowing how much is allocated and how much is used, per allocator type, would help. You can also run `rabbitmq-diagnostics observer`, go to the ets table page (`E` + Enter), sort by memory (`m` + Enter) and provide a screenshot of what you see.

**Reply by michaelklishin** [reply]
Yup, this definitely looks like a memory fragmentation problem. There is quite a number of things that [can be tweaked](https://www.rabbitmq.com/runtime.html#allocators).

**Reply by kbudde** [reply]
I checked the rabbitmq-diagnostics but it looks "normal" to me and much less than expected.
![image](https://user-images.githubusercontent.com/1072181/124008188-71c99180-d9dc-11eb-8c60-684f1baebd1d.png)

Metrics from the same time:
![image](https://user-images.githubusercontent.com/1072181/124008467-bbb27780-d9dc-11eb-8e7d-bf8dd9bf71f1.png).

What would you recommend? 
IMO I would scale up to three nodes with HA-sync exactly/2 and test with different settings on 1 node for comparison or is this something you would not recommend?

**Reply by lhoguin** [reply]
The `mqtt_node` table is owned by `ra_coordination_log_ets`. I do not know if it is expected that it is that large. @kjnilsson might know.

Otherwise, provide an answer for "Knowing how much is allocated and how much is used, per allocator type, would help". The Erlang Memory Allocators dashboard has all the info we need. The bits you have showed so far are too general to know anything other than there's a lot of unused memory.

**Reply by kbudde** [reply]
@lhoguin Just say me what you need and I'll do my best.
Here are the values from ets_alloc
![image](https://user-images.githubusercontent.com/1072181/124118309-7a1fdc00-da71-11eb-8c45-2ee3efab93cc.png)

just in case:
binary_alloc
![image](https://user-images.githubusercontent.com/1072181/124118457-a63b5d00-da71-11eb-9509-c356a35878c8.png)
eheap
![image](https://user-images.githubusercontent.com/1072181/124118787-03cfa980-da72-11eb-870d-ef054aad8e27.png)

**Reply by lhoguin** [reply]
It doesn't look too bad but there's room for improvement. I'm a bit surprised to see around 70% utlisation because the current default flags ought to reclaim memory more than they seem to do for you. Did you change the VM flags?

You could check observer again only this time sorting by `m` without changing page to see if a process uses more memory than others.

**Reply by kbudde** [reply]
VM Flags are still unchanged. 
![image](https://user-images.githubusercontent.com/1072181/124121607-61192a00-da75-11eb-89b3-066ecf5ce537.png)

![image](https://user-images.githubusercontent.com/1072181/124121487-3af38a00-da75-11eb-9c9d-1ebece638e36.png)

**Reply by lhoguin** [reply]
The processes using the most memory are metrics related. This is typical, but one of them has almost 200MB usage on your screenshot, which is a bit much. It would be worth keeping track of this view over time to see if this process memory usage increases further.

There is also an HTTP request process that has 28MB. That's a lot. It might be the prometheus /metrics endpoint; or the per-object endpoint; or it could be some other script you have configured to periodically check something in RabbitMQ. It would be good to figure it out. You can view the process details from that view by pressing the number of the process (on the left) + Enter, and then print the screen that appears, press `C` to get the stacktrace screen and print it as well. As this is typically a short lived process this might be difficult to get however.

**Reply by kbudde** [reply]
Ok, we decided to reboot it today. The daily peak memory usage was already quite high and we did not want to risk reaching high memory watermark over the weekend.

The http request is clear, we have still the rabbitmq_exporter running which is querying the management api. We doubled the scrape intervall to reduce the load here. Unfortunately it's still needed.

After reboot it looks less stressed:
![image](https://user-images.githubusercontent.com/1072181/124173328-3053e780-daab-11eb-8299-00cb4f3400d3.png)
![image](https://user-images.githubusercontent.com/1072181/124173884-f46d5200-daab-11eb-9a0f-57feaad864f9.png)
![image](https://user-images.githubusercontent.com/1072181/124173914-fe8f5080-daab-11eb-8a67-25cf12447725.png)
![image](https://user-images.githubusercontent.com/1072181/124173919-00591400-daac-11eb-85e7-0eef1b375c8d.png)

**Reply by lhoguin** [reply]
OK. I don't think this is fragmentation in any case. Would be good to see current ets tables, see what size the big table currently has. Probably not nearly as big considering your numbers.

Would be good to keep an eye on how fast memory grows to see if doubling the scrape interval helped.

Not sure what else to suggest at this point.

**Reply by kbudde** [reply]
![image](https://user-images.githubusercontent.com/1072181/124233982-bd328b80-db13-11eb-8fd2-c4211102fb44.png)

**Reply by michaelklishin** [reply]
Nothing around those tables particularly stands out. Without a way to reproduce your workload — specifically how/when connections are opened and closed — we can only guess whether the `mqtt_node` table is actually large or expected to be of this size.

See [Dealing with High Connection Churn](https://www.rabbitmq.com/networking.html#dealing-with-high-connection-churn) if you have connection churn. It can be that the MQTT client ID tracker still leaks the tiny records it uses. It can also be that your have a ton of connections in `TIME_WAIT` that seem open to the runtime and thus RabbitMQ, which means
that data cannot (yet) be deleted by the MQTT plugin.

**Reply by kbudde** [reply]
hi @michaelklishin, I don't think the issue is high connection churn here.
At peak time we have normally around around one connect/disconnect per second. Time_waits are not going up.

But I suspect it might be related to client ID. 
The client ID is not stable at our side. It's partly random; so the number of different client IDs is theoretically not limited. If the mqtt plugin keeps statistics on them, I suspect it might be normal to see this grow. Is there anything I can tune here? Or maybe something I can check?

I will try to reproduce the issue on our test system.

**Reply by kbudde** [reply]
Just as a small update I tried to reproduce the issue but without success so far.
I'm unsure how to proceed.

**Comment by kbudde** [comment]
Hi,

I've noticed something else which might be related to the unusual memory consumption.

According to the RabbitMQ-Overview Dashboard: We have currently around 30 msg/s delivered and no messages are waiting; so all queues are nearly empty.
But the metrics for "messages published" and "messages routed to queues" are showing around 200 times higher message rates. 
![image](https://user-images.githubusercontent.com/1072181/124233639-54e3aa00-db13-11eb-9427-2d96cc792f50.png)
![image](https://user-images.githubusercontent.com/1072181/124233741-6d53c480-db13-11eb-9e68-68a0daf2e6fc.png)

**Reply by michaelklishin** [reply]
This is more likely to be an issue with Prometheus metrics or their aggregation than anything related to MQTT.
